File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.java
class Solution {
  public int maxProfit(int[] prices) {
    int ans = 0;
    int[] fw = new int[prices.length];
    fw[0] = 0;
    int[] bw = new int[prices.length];
    bw[bw.length - 1] = 0;
    int lMin = prices[0];
    int rMax = prices[prices.length - 1];
    for (int i = 1; i < prices.length; i++) {
      fw[i] = Math.max(fw[i - 1], prices[i] - lMin);
      lMin = Math.min(lMin, prices[i]);
    }
    for (int i = prices.length - 2; i >= 0; i--) {
      bw[i] = Math.max(bw[i + 1], rMax - prices[i]);
      rMax = Math.max(rMax, prices[i]);
    }
    for (int i = 0; i < prices.length; i++) {
      ans = Math.max(ans, fw[i] + bw[i]);
    }
    return ans;
  }
}

File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.py
class Solution:
  def maxProfit(self, prices: List[int]) -> int:
    ans = 0
    fw, bw = [0]*len(prices), [0]*len(prices)
    lMin, rMax = prices[0], prices[-1]
    for i in range(1, len(prices)):
      fw[i] = max(fw[i-1], prices[i]-lMin)
      lMin = min(lMin, prices[i])
    for i in range(len(prices)-2, -1, -1):
      bw[i] = max(bw[i+1], rMax-prices[i])
      rMax = max(rMax, prices[i])
    for i in range(len(prices)):
      ans = max(ans, fw[i]+bw[i])
    return ans